draft-ietf-quic-version-negotiation.md

```diff
@@ -387,18 +387,37 @@ based on information received from a Version Negotiation packet MUST ignore any
 Version Negotiation packets it receives in response to that connection attempt.
 
 Both endpoints MUST parse their peer's Version Information during the handshake.
-If parsing the Version Information failed (for example, if it is too short or if
+If that leads to a parsing failure (for example, if it is too short or if
 its length is not divisible by four), then the endpoint MUST close the
 connection; if the connection was using QUIC version 1, that connection closure
 MUST use a transport error of type TRANSPORT_PARAMETER_ERROR. If an endpoint
 receives a Chosen Version equal to zero, or any Available Version equal to zero,
-it MUST treat it as a parsing failure.
+it MUST treat it as a parsing failure. If a server receives a Version
+Information where the Chosen Version is not included in Available Versions, it
+MUST treat it as a parsing failure.
 
 Every QUIC version that supports version negotiation MUST define a method for
 closing the connection with a version negotiation error. For QUIC version 1,
 version negotiation errors are signaled using a transport error of type
 VERSION_NEGOTIATION_ERROR; see {{iana-error}}.
 
+When a server receives a client's first flight, the server will first establish
+which QUIC version is in use for this connection in order to properly parse the
+first flight. For example, the server determines that QUIC version 1 is in use
+by observing that the Version field of the first Long Header packet it receives
+is set to 0x00000001. When the server then parses the client's Version
+Information, the server MUST validate that the client's Chosen Version matches
+the version in use for the connection. If the two differ, the server MUST close
```

Comment:
I think that the problem I'm seeing here is rooted in the word "parses" and how it connects to the "in use" bit.  I think that you might want to use "processes" in the first instance.  For the second, use the version of the Initial packet that was used to open the connection rather than a potentially over-broad "in use".

Maybe instead:

> If the server processes Version Information from the client and the Chosen Version differs from the version of the packet that initiated the connection attempt, it MUST close the connection with a version negotiation error.

However, this implies something about Retry packets, so I get why you went with "in use" to avoid that.  Maybe instead:


> If the server processes Version Information from the client and the Chosen Version differs from the version from the version of packets the server has received for this connection attempt, it MUST close the connection with a version negotiation error.

That is, rely on an expectation that the client uses a single, consistent version for its connection attempt.